src/ufuncs.c

```diff
@@ -722,6 +734,28 @@ static void haussdorf_distance_densify_func(char **args, npy_intp *dimensions,
 static PyUFuncGenericFunction haussdorf_distance_densify_funcs[1] = {&haussdorf_distance_densify_func};
 
 
+static char frechet_distance_densify_dtypes[4] = {NPY_OBJECT, NPY_OBJECT, NPY_DOUBLE, NPY_DOUBLE};
+static void frechet_distance_densify_func(char **args, npy_intp *dimensions,
+                                            npy_intp *steps, void *data)
+{
+    void *context_handle = geos_context[0];
+    GEOSGeometry *in1, *in2;
+
+    TERNARY_LOOP {
+        /* get the geometries: return on error */
+        if (!get_geom(*(GeometryObject **)ip1, &in1)) { return; }
+        if (!get_geom(*(GeometryObject **)ip2, &in2)) { return; }
+        double in3 = *(double *) ip3;
+        if ((in1 == NULL) | (in2 == NULL) | npy_isnan(in3) | GEOSisEmpty_r(context_handle, in1) | GEOSisEmpty_r(context_handle, in2)) {
```

Comment:
If `densify` is negative, we get this:

```
pygeos.GEOSException: IllegalArgumentException: Fraction is not in range (0.0 - 1.0]
```

Personally, I think we should specifically check for this case in Python / C and raise a `ValueError` instead.  (and add a corresponding test to validate it is raised)

Reply:
Sure, I think we can check if the value is specifically in the range (0.0 - 1.0] in Python, and raise a ValueError if not.

I have just pushed a commit for the same. I also excluded nan from valid values of `densify` in this commit. @caspervdw if you have any reason why we should allow `nan`, please let me know I will adjust the code accordingly. Otherwise, @brendan-ward I think we can go ahead with any other suggestions or for a merge.

Reply:
The reason for allowing nan in `haussdorf_distance_densify` is that GEOS handles that situation correctly: it returns `nan`.

Whenever GEOS handles the situation correctly, there is no reason to put up another Python layer of exception handling.

Reply:
@caspervdw GEOS isn't handling `np.nan` handled specifically in the C code (i.e., above) - which means it wasn't a choice made inside GEOS, right?

Personally, I think while there are a number of valid cases where the inputs can be a mix of `nan` and valid values, it feels like a validation problem for parameters that control the output of the function - and are parameters that are unlikely to vary based on your input geometries.

Example: I can see cases where I could pass in an array of `densify` values directly or via broadcasting; maybe some geometries need to be densified more than others.  I'm not deriving `densify` at runtime from something else that may return `nan` (though in theory someone could; e.g., as a function fo distance), so it means that I goofed up if `nan`s are in the list, and I'd rather have to handle that specifically rather than silently getting `nan` in my output.

For example, numpy doesn't let me get away with it for parameters that control the output in a roughly similar way:
```
>>> np.repeat([1], [np.nan])
...
ValueError: cannot convert float NaN to integer
```

But - these are bigger issues we should probably take up elsewhere...

If your sense is that `nan` is a valid input in this case and the output should deterministically be `nan`, instead of indicating a validation problem, then my apologies to @chaitan94 for suggesting otherwise!


As for exceptions, I realize this is very much my own opinion, but my suggestion is that `GEOSException` indicates that something went wrong inside the GEOS function when we otherwise expected it to succeed.  There are a number of places in C that we're returning `ValueError` for invalid inputs (also `TypeError`), and that seems like a reasonable pattern to follow elsewhere: it indicates we deliberately looked for invalid values and handled them before ever calling into GEOS.  The text of the GEOS error in this particular case is sufficiently clear - but I'm not sure we can always count on GEOS to give us back meaningful exceptions for data validation issues.

Reply:
Got it, I made changes to return NaN when densify is passed as NaN.
I have also handled ndarray case of densify and added a test for the same.

Apart from that, I see two options as of now:

1. Handle values <= 0 and >1 ourselves and raise ValueError
2. Just rely on GEOS exception, add tests for that directly

I pushed a commit for the implementation of option 1.

For option 2, we won't have an additional check in the C code, and we will instead do this in the tests:

```py
with pytest.raises(pygeos.GEOSException, match=r"IllegalArgumentException: Fraction is not in range \(0.0 - 1.0]"):
    actual = pygeos.frechet_distance(line_string, line_string, densify=densify)
```

Pros of option 1: We can add more information like for example when I did with `got 2 instead`.
Pros of option 2: No need for additional code and redundant checks in pygeos and it will stay in sync with GEOS going forward as well.

Based on your input I can add another commit to change the implementation to option 2 instead.

Reply:
@chaitan94 thanks for making the updates!  And apologies for introducing confusion where we still have some different ideas about how to approach validation / error handling.  We'll get there!

👍 on your option 1.  In my personal opinion, as a general rule, it seems risky to rely on the exception messages from GEOS as always fitting our usage here; while the code level dependency on GEOS for implementation is unavoidable, the dependency on exception messages and validation doesn't have to be.

That said, we could go with option 2 as a general rule throughout the codebase, so long as we are diligent about adding tests for those exceptions, and manually verify that the exception we get back from GEOS is good enough (i.e., via the `match` parameter as you've suggested in 2)

Reply:
Sorry for catching on again so late. Thanks for the thorough discussion on this subject already.

> @caspervdw GEOS isn't handling np.nan handled specifically in the C code (i.e., above) - which means it wasn't a choice made inside GEOS, right?

You are correct, sorry, I didn't actually check.

> Personally, I think while there are a number of valid cases where the inputs can be a mix of nan and valid values, it feels like a validation problem for parameters that control the output of the function - and are parameters that are unlikely to vary based on your input geometries.

Using broadcasting is a general design pattern in PyGEOS. We sometimes diverge from it (see the `buffer` implementation), but that's because of technical difficulties on the GOES CAPI side. .

Propagating missing is also a general design pattern in PyGEOS. For Geometries, `None` indicates "missing" , for floats `NaN` does it. For integers and booleans, there is no value to indicate "missing".

The two points distinguished by @chaitan94 do need some more discussion.

> 1. Handle values <= 0 and >1 ourselves and raise ValueError
> 2. Just rely on GEOS exception, add tests for that directly

My preference is: if it is done by GEOS, PyGEOS should not do it again. More code means more maintenance, more tests, more bugs, etc. We need focus.

>  +1 on your option 1. In my personal opinion, as a general rule, it seems risky to rely on the   exception messages from GEOS as always fitting our usage here; while the code level dependency on GEOS for implementation is unavoidable, the dependency on exception messages and validation doesn't have to be.

So that is the exact opposite. I'd like option 2. Discussion is good! Let's continue this in a separate issue. For now, leave it as is, I'll approve so this can be merged.

On a side note, please don't mistake my shortish replies as a lack of appreciation of the work you are doing on this PR. I really appreciate the contributions to PyGEOS.